server/src/main/java/org/opensearch/common/time/JavaDateFormatter.java

```diff
@@ -237,6 +281,12 @@ private TemporalAccessor doParse(String input) {
                 ParsePosition pos = new ParsePosition(0);
                 Object object = formatter.toFormat().parseObject(input, pos);
                 if (parsingSucceeded(object, input, pos)) {
+                    if (FeatureFlags.isEnabled(FeatureFlags.DATETIME_FORMATTER_CACHING)
```

Comment:
@dblock do we really need feature flag here? We have setting already with default value set to `false`, the feature flag look redundant

Reply:
Feature flag was used to give the users a knob to turn it off incase they face any regression. Lmk your thoughts as well on this

Reply:
A bit confused now now, I see:
- DATETIME_FORMATTER_CACHING_SETTING, a setting to turn it on  / off 
- and a feature flag DATETIME_FORMATTER_CACHING to do the same thing?
I think having only the setting should be sufficient, it is the same on / off knob

Reply:
Yes, so this follows the existing convention we have for featureflags, we've a name of feature flag which is usually defined as a setting in the code whereas the actual setting name differs and is defined without setting in variable name. You can see this being followed in other features like telemetry, segment replication, concurrent segment search etc.

>I think DATETIME_FORMATTER_CACHING  is just the setting name(string) for DATETIME_FORMATTER_CACHING_SETTING 

Thanks ankit, this statement explains it perfectly.

>We don't need a FeatureFlag here. We need to move this to cluster/index setting.
Also, i think we should keep the setting dynamic and default as true

On this one, we need to think that if we go down the setting path then will we be needing setting for every perf optimization we will be doing? It'll result in lots of settings to be maintained. Featureflags have an advantage that we can remove them after 1 or more minor versions, once the feature gets battle tested in real world usage. I'm of the opinion that we shouldn't even be having either feature flag or any setting for perf optimizations but just to be on the safe side i.e. for any workloads which we can't test for it may result in regression we can have a featureflag to turn it off. 

Would like to know your thoughts on this @ankitkala , @reta and @dblock

Reply:
> Would like to know your thoughts on this @ankitkala , @reta and @dblock

For some reason I missed the fact that the feature flag (even being just a string) are defined through settings, all good here, thanks @ankitkala (sorry for confusion @CaptainDredge ).

> On this one, we need to think that if we go down the setting path then will we be needing setting for every perf optimization we will be doing?

For this specific feature I would also prefer to see it as a feature flag and not a setting (in general, we should either apply it as the default path or not).

Reply:
>For this specific feature I would also prefer to see it as a feature flag and not a setting (in general, we should either apply it as the default path or not).

Yes I'm also of the opinion that it should be part of the default path. @reta can I get your thorough review on the changes? Would like to address any more concerns or suggestions

Reply:
I would just do it and be confident that it will work by doing more testing. If we're not confident we should not do it :) The point of feature flags is to let users test _changing external behavior_, but this is a pure performance optimization, so I don't see anyone not wanting it.

Reply:
> I would just do it and be confident that it will work by doing more testing. If we're not confident we should not do it :) 

I think we could make it `true` by default but still let the users to switch it off (at least in during one release). The change looks like only optimization but it is not - quite a few changes on the side of preserving the existing formatting behaviour.

Reply:
+1 to keep it `true` by default and using the setting as a safety control.

Reply:
Since, the quorum is on keeping the feature flag, I've changed the default as `true`